### PR TITLE
Use RtlDllShutdownInProgress to detect process shutdown on Windows

### LINKED ISSUE #103624: Test failure: CoreMangLib/system/span/RefStructWithSpan/RefStructWithSpan.cmd
**Failed in:** [runtime-coreclr outerloop 20240617.4](https://dev.azure.com/dnceng-public/public/_build/results?buildId=711071&view=ms.vss-test-web.build-test-results-tab&runId=17771968&resultId=122382&paneView=debug)

**Failed tests:**
```
coreclr windows arm64 Checked no_tiered_compilation @ Windows.11.Arm64.Open
    - CoreMangLib/system/span/RefStructWithSpan/RefStructWithSpan.cmd
```

**Error message:**
```
 
cmdLine:C:hwB88F09C1wC2F00A46eCoreMangLibCoreMangLib../system/span/RefStructWithSpan/RefStructWithSpan.cmd Timed Out (timeout in milliseconds: 600000 from variable __TestTimeout, start: 6/18/2024 2:47:58 AM, end: 6/18/2024 2:57:58 AM)

Return code:      -100
Raw output file:      C:hwB88F09C1wC2F00A46uploadssystemspanRefStructWithSpanoutput.txt
Raw output:
BEGIN EXECUTION
 "C:hwB88F09C1pcorerun.exe" -p "System.Reflection.Metadata.MetadataUpdater.IsSupported=false" -p "System.Runtime.Serialization.EnableUnsafeBinaryFormatterSerialization=true"  RefStructWithSpan.dll 

cmdLine:C:hwB88F09C1wC2F00A46eCoreMangLibCoreMangLib../system/span/RefStructWithSpan/RefStructWithSpan.cmd Timed Out (timeout in milliseconds: 600000 from variable __TestTimeout, start: 6/18/2024 2:47:58 AM, end: 6/18/2024 2:57:58 AM)
Test Harness Exitcode is : -100
To run the test:
Set up CORE_ROOT and run.
> C:hwB88F09C1wC2F00A46eCoreMangLibCoreMangLib../system/span/RefStructWithSpan/RefStructWithSpan.cmd
```

**Stack trace:**
```
   at Xunit.Assert.True(Nullable`1 condition, String userMessage) in /_/src/Microsoft.DotNet.XUnitAssert/src/BooleanAsserts.cs:line 146
   at TestLibrary.OutOfProcessTest.RunOutOfProcessTest(String assemblyPath, String testPathPrefix)
   at Program.<<Main>$>g__TestExecutor46|0_47(StreamWriter tempLogSw, StreamWriter statsCsvSw, <>c__DisplayClass0_0&)
```

### PR DESCRIPTION
**Pull request by jkotas**
Switching to cooperative mode is not safe during process shutdown on Windows. Process shutdown can terminate a thread in the middle of the GC. The shutdown thread deadlocks if it tries to switch to cooperative mode and wait for the GC to finish in this situation.

Use RtlDllShutdownInProgress Windows API to detect process shutdown to skip cleanup that has to be done in cooperative mode.

The existing g_fProcessDetach flag is set too late - using this flag to skip cooperative mode switch would lead to shutdown deadlocks, and the existing g_fEEShutDown flag is set too early - using this flag to skip cooperative mode switch would lead to shutdown crashes.

Fixes https://github.com/dotnet/runtime/issues/103624

### CODE REVIEW

**Comment by jkotas** [comment]
This is not 100% reliable fix, but it matches what we do in similar situations during shutdown currently.

**Review by jkoritzinsky — APPROVED** [review]

**Comment by jkotas** [comment]
Unfortunately, this fix does not work well. It replaces intermittent hang with intermittent crash during shutdown.

**Comment by jkotas** [comment]
> What if we do the alloc context fixing during the destruction of the RuntimeThreadLocals object? 

I do not see how it would help.

This is the sequence that leads to the hang:
- OS initiates process shutdown
- There is a thread in the middle of the GC that gets killed by the process shutdown sequence. It means that the process state is indefinitely stuck in "GC in progress" state
- OS sends thread detach notification to the last thread in the process that runs the thread-local destructors. We try to switch to cooperative mode and wait for GC to finish - but that's never going to happen.

The problem with the current fix is that it is starts skipping the cleanup too early while there are still multiple threads running in the process. It means that we can end up running a GC without cleaned up alloc context that the GC gets very unhappy about.

I have been looking for a more precise detection of the moment after which it is fine to start skipping the cleanup.

**Comment by jkotas** [comment]
Ok, I think I found something that works. I have also fixed other instances of thread cleanup switching to cooperative mode that I found during testing and review.

File: src/coreclr/vm/vars.hpp [hunk 1]
```diff
@@ -459,12 +459,24 @@ GVAL_DECL(bool, g_metadataUpdatesApplied);
...
+inline bool IsAtProcessExit()
 {
     SUPPORTS_DAC;
+#if defined(DACCESS_COMPILE) || !defined(HOST_WINDOWS)
     return g_fProcessDetach;
+#else
+    // RtlDllShutdownInProgress provides more accurace information about whether the process is shutting down.
+    // Use it if it is available to avoid shutdown deadlocks.
+    // https://learn.microsoft.com/windows/win32/devnotes/rtldllshutdowninprogress
+    return g_pfnRtlDllShutdownInProgress();
```
Comment (MichalPetryka) [inline src/coreclr/vm/vars.hpp:478]: Shouldn't this handle the case of it not being present?

Comment (jkotas) [reply src/coreclr/vm/vars.hpp:478]: It is handled here: https://github.com/dotnet/runtime/pull/103877/files#diff-63935b2e5a324d25feff725f83200eaa716c69fdef7529d2bd3cfa35620105aeR399-R403

**Review by davidwrighton — APPROVED** [review]
So, this is race condition city, and I think that there are still races that exist. Notably, I don't understand when the GC thread can be eliminated without running its cleanup logic, so I don't know how to analyze this for complete correctness, but based on the description you provided this should strictly reduce the set of deadlocks that are likely to occur.

**Comment by jkotas** [comment]
> I don't understand when the GC thread can be eliminated without running its cleanup logic

Any shutdown with ongoing background activity is prone to this problem. What happens is:
- Some thread (typically the main app thread) decides to shutdown the process. It must switch to preemptive mode to do so.
- Some background thread starts GC
- Windows kernel suspends all threads (except the one thread running the shutdown) and kills them. The thread running in the middle of the GC can be one of those.
- The one remaining shutdown thread is resumed to proceed with the cleanup. If it ever tries to switch to cooperative mode and wait for GC to finish, it is going to hang.

> I think that there are still races that exist.

Yes, I agree. The code has a lot of checks that should be either unnecessary or that have race conditions.

File: src/coreclr/vm/vars.hpp [hunk 2]
```diff
@@ -459,12 +459,24 @@ GVAL_DECL(bool, g_metadataUpdatesApplied);
...
 // Indicates whether we're executing shut down as a result of DllMain
 // (DLL_PROCESS_DETACH). See comments at code:EEShutDown for details.
-inline BOOL    IsAtProcessExit()
+inline bool IsAtProcessExit()
 {
     SUPPORTS_DAC;
+#if defined(DACCESS_COMPILE) || !defined(HOST_WINDOWS)
     return g_fProcessDetach;
+#else
+    // RtlDllShutdownInProgress provides more accurace information about whether the process is shutting down.
```
Comment (Hrxn) [inline src/coreclr/vm/vars.hpp:475]: `accurace` -> `accurate` ?

Comment (jkotas) [reply src/coreclr/vm/vars.hpp:475]: Thanks - fix included in https://github.com/dotnet/runtime/pull/104318/

**Pull request merged**